### ISSUE TITLE
Distributions are not nn.Modules

### ISSUE THREAD

**Issue by fritzo** [pytorch/pytorch#31009]
## 🚀 Feature
Many users are surprised that `Distribution` objects are not `nn.Module`s, and that updating input parameters does not change a distribution. This issue aims to explain the current state of affairs, explain challenges to making `Distributions` into `nn.Module`s, and collect ideas that might help with such a refactoring.

## Motivation

PyTorch users often try to optimize parameters of `Distribution` instances as if they were `nn.Module`s:
```py
loc = nn.Parameter(torch.tensor(0.))
log_scale = nn.Parameter(torch.tensor(1.).log())
n = dist.Normal(loc, log_scale.exp())
n.log_prob(data).sum().backward()
optim.step([loc, log_scale])
```

## Pitch

Many users expect `Distribution` to be a subclass of `nn.Module`:
```py
class Distribution(nn.Module):
    ...
class Normal(Distribution):
   ...
```

## Why are Distributions not currently nn.Modules?

The original [design doc](https://docs.google.com/document/d/1wnABg0cdyaVMr-Xqz_brHBwnPJuzTeSk3Oqko2HHSfE/edit?usp=sharing) focused on batch semantics, parameter constraints, and transforms. The authors of that early vision collectively decided that it would be simpler if distributions were immutable, thereby permitting caching of intermediate results, and simplifying code. In this sense, `torch.distributions` uses immutable flyweight objects to bundle together multiple pure functions as methods, sort of an object-oriented version of `scipy.stats`. Additionally the authors aimed to closely follow the Tensorflow Probability interface so as to make it easy to port models between PyTorch and Tensorflow.

## Challenges to making Distributions into nn.Modules

There are now many distributions and lots of mathematically intricate code relying on some challenging assumptions:

### Challenge 1: constrained parameters

Distribution parameters follow standard statistical conventions, aiming to agree with Tensorflow or scipy.stats where possible. Standard parameters are often *constrained*, for example `Normal` is parameterized by an unconstrained `loc` tensor and a positively constrained `scale` tensor.

It is unclear how to best support optimization of constrained parameters in idiomatic PyTorch. One possibility is the support for constrained parameters in [PyroModule](http://docs.pyro.ai/en/1.1.0/nn.html#module-pyro.nn.module) introduced in Pyro 1.0; we could conceivable move this upstream to PyTorch as a `ConstrainedModule`.

### Challenge 2: cached intermediate computations

A number of distributions use `@lazy_property` to cache multiple intermediate computations, and to store unbroadcasted versions of tensors. This is especially important in variational inference applications where a single step includes calls to both `.rsample()` and `.log_prob()`, and intermediate computations can be shared. For example see the highly optimized [MultivariateNormal](https://github.com/pytorch/pytorch/blob/master/torch/distributions/multivariate_normal.py) implementation.

One possibility to implement caching is to create a temporary cache with the scope of a single `.__call__()` invocation. This pattern is partially used in [PyroModule]() to cache sample statements; however this would require annotating each user-facing method with a hook-enabling decorator, since not all methods would call `.__call__()`/`.forward()` (see [@pyro_method](http://docs.pyro.ai/en/1.1.0/nn.html#pyro.nn.module.pyro_method) for the approach taken in `PyroModule`).

### Challenge 3: alternate parameterizations

The original `torch.Distributions` design followed Tensorflow by opting for single distributions with multiple alternate parameterizations specified by kwargs, e.g. `Bernoulli(probs=probs)` versus `Bernoulli(logits=logits)`. While this simplifies code initially, it complicates the handling of mutable parameters (e.g. is `probs` a function of `logits` or vice versa?).

One solution (followed in [NumPyro](https://numpyro.readthedocs.io/en/latest/distributions.html#discrete-distributions) and [Funsor](https://funsor.readthedocs.io/en/stable/distributions.html)) is to retract that initial decision and split into two multiple classes
```py
class BernoulliLogits(...): ...
class BernoulliProbs(...): ...
def Bernoulli(probs=None, logits=None):
    if probs is not None:
        return BernoulliProbs(probs)
    if logits is not None:
        return BernoulliLogits(logits)
```
however this change is backwards-incompatible: many users subclass `Bernoulli` and other distribution classes.

### Challenge 4: jit support

Distributions currently provide jit.trace support by unwrapping to a sequence of PyTorch ops, where a single trace might include (1) construction via `.__init__()`, (2) evaluation of `.rsample()`, and (3) evaluation of `.log_prob()`. It is unclear how to provide this in the `nn.Module` interface, where jitting prefers to call a single `.forward()` method. (@t-vi feel free to edit if you see other issues!)

cc @vincentqb @fritzo @neerajprad @fehiepsi @alicanb @vishwakftw @t-vi

**Comment by t-vi** [comment]
I'd add JIT support to the constraints.
I'm still very vaguely hopeful that introducing a calculated param extension like #7313 might help with the constraints and caching. (Interaction of spectral norm / weight norm with hooks and JIT came up recently, too.)
Unfortunately, barring unforeseen interest, I'm not going to be working on it.

**Comment by vadimkantorov** [comment]
About reparameterizations, there is also this proposal: https://github.com/pytorch/pytorch/issues/28937

If a big refactoring is being thought through, maybe deisgn for invertible ops can also get into core modules: https://github.com/pytorch/pytorch/issues/23756. Also I was wondering, whether it's possible to use torch.log instead of LogTransform etc. Probably some sort of lazy evaluation / standard PyTorch op polymorphism is needed?

**Comment by KeAWang** [comment]
It would also be nice to have `torch.distributions.Transform`s also turned into `nn.Module`s. That way if I have a `TransformedDistribution`, I can move all of its parameters/buffers to a certain device with a simple call to `.to`. Right now I would have to initialize a new `TransformedDistribution` while passing in all the newly moved parameters for each of the `Transform`s

**Comment by fritzo** [comment]
@KeAWang good point, @stefanwebb has already refactored some of the Pyro codebase to use [TransformModule](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/torch_transform.py#L7) and is using it in his normalizing flows and transforms [module](http://docs.pyro.ai/en/stable/distributions.html#transformmodules).

**Comment by geoffsmith** [comment]
Any suggestions for a workaround for this that we could use in the meantime?

**Comment by fritzo** [comment]
@geoffsmith I would recommend constructing distributions lazily, as-needed.

Closing this as infeasible.
